packages/http/server/response_builder.mare

```diff
@@ -0,0 +1,86 @@
+:import "bytes" (ByteStreamWritable)
+
+:class ResponseBuilder
+  :let _res ByteStreamWritable
+
+  :new (target ByteStreamWritable)
+    @_res = target
+    @_res << b"HTTP/1.1 " // TODO: I don't like inserting the version here, but carry on...
```

Comment:
Maybe we should just include the version in the status line?

Reply:
For now I'd rather leave it here. Putting that in the status definition made the lines too long xD.

I imagine a second iteration of this where we allow the users to call status, header and body in any order, and then `finish` puts everything together.

Reply:
The thing about that is that it would decrease performance a bit, due to the extra operations going on. One of the reasons I suggested this design was because I think it will have the best performance possible, but that's something we can measure and see what the impact is.

There's a trick we can do with the type system to prevent accidentally calling them out of order, so we can ensure there's no mistakes on the caller side that way.

For cases where you actually want to call them out of order, we could potentially have a different workflow where you create a `Response` object, then have `ResponseBuilder` able to accept the entire `Response` object and write it all at once. Maybe in that case it makes sense to call this `ResponseWriter` rather than `ResponseBuilder`? I'm not sure - I could probably be convinced either way.

Reply:
>The thing about that is that it would decrease performance a bit, due to the extra operations going on. One of the reasons I suggested this design was because I think it will have the best performance possible, but that's something we can measure and see what the impact is.

Definitely makes sense for better performance. I would love to benchmark it against the more flexible approach you described here:

>For cases where you actually want to call them out of order, we could potentially have a different workflow where you create a Response object, then have ResponseBuilder able to accept the entire Response object and write it all at once. Maybe in that case it makes sense to call this ResponseWriter rather than ResponseBuilder? I'm not sure - I could probably be convinced either way.

I like the idea of not penalizing performance but compared with other web frameworks in the wild I think users would like a less restrictive approach to build responses. If we make our `Builder` a `Writer`, and then allow for a `Response` object, folks could use either approach depending on their needs. In order to decide if this flexibility is good or bad I would like to see how much performance is affected.

For now it makes sense to keep the existing approach and once we have rounded off a few corners we can put together a `Response` object to compare against. What do you think?

>There's a trick we can do with the type system to prevent accidentally calling them out of order, so we can ensure there's no mistakes on the caller side that way.

Oh? Interesting, would love to know more.